sam.c

```diff
@@ -4614,23 +4625,26 @@ uint8_t *bam_aux_get(const bam1_t *b, const char tag[2])
     return NULL;
 }
 
-// s MUST BE returned by bam_aux_get()
 int bam_aux_del(bam1_t *b, uint8_t *s)
 {
-    uint8_t *p, *aux;
-    int l_aux = bam_get_l_aux(b);
-    aux = bam_get_aux(b);
-    p = s - 2;
-    s = skip_aux(s, aux + l_aux);
-    if (s == NULL) goto bad_aux;
-    memmove(p, s, l_aux - (s - aux));
-    b->l_data -= s - p;
-    return 0;
+    s = bam_aux_remove(b, s);
+    return (s || errno == ENOENT)? 0 : -1;
+}
+
+uint8_t *bam_aux_remove(bam1_t *b, uint8_t *s)
+{
+    uint8_t *end = b->data + b->l_data;
+    uint8_t *next = skip_aux(s, end);
+    if (next == NULL) goto bad_aux;
+    memmove(s-2, next, end - next);
+    b->l_data -= next - (s-2);
+    if (next >= end) { errno = ENOENT; return NULL; }
```

Comment:
Nit-picking: If `next > end` then the memmove has attempting moving a negative size.  If `next == end` then memmove is a no-op.  Both of these imply memmove should be after this check.

In reality though `next <= end` in all scenarios due to the error checking in `skip_aux`, so it just *looks* wrong rather than actually being wrong.

Reply:
Good spotting — this indicates a bug, probably introduced during that refactoring!

That `if` is supposed to return NULL or `s` depending on whether `s` was previously the last entry. Probably `end` needs to be recalculated now that `l_data` has changed, or something equivalent.

Reply:
I think this was not a bug in the end, because as you say `next > end` can't happen. But I have reorganised the code and think it's probably clearer now anyway.